### PR TITLE
retry creating dnstest.Server if tcp or udp port is in use

### PR DESCRIPTION
**Pull request by rdrozhdzh**
### 1. What does this pull request do?
retry creating tcp and udp listeners if desired UDP port is already in use

### 2. Which issues (if any) are related?
#1785

### 3. Which documentation changes (if any) need to be made?
none

### CODE REVIEW

**Comment by miekg** [comment]
The doing tcp first should be sufficient, complicating this further doesn't add much IMO

**Comment by rdrozhdzh** [comment]
@miekg, I reopened #1785 because I caught the test failed with panic due to this issue, see https://github.com/coredns/coredns/issues/1785#issuecomment-389111894

**Comment by miekg** [comment]
none of the tests assume the testserver will fail, so this is a pretty harmless panic

**Comment by rdrozhdzh** [comment]
reopening since the issue is getting annoying

File: plugin/pkg/dnstest/server.go [hunk 1]
```diff
@@ -23,18 +23,28 @@ func NewServer(f dns.HandlerFunc) *Server {
 	ch1 := make(chan bool)
 	ch2 := make(chan bool)
 
-	l, _ := net.Listen("tcp", ":0")
-	if l == nil {
-		return nil
+	s1 := &dns.Server{} // udp
+	s2 := &dns.Server{} // tcp
+
+	for i := 0; i < 1000; i++ { // 1000 attempts
```
Comment (miekg) [inline plugin/pkg/dnstest/server.go]: Can we make it, 5 ? 1000 looks like very much an overkill?

File: plugin/pkg/dnstest/server.go [hunk 2]
```diff
@@ -23,18 +23,28 @@ func NewServer(f dns.HandlerFunc) *Server {
...
+		s2.Listener, _ = net.Listen("tcp", ":0")
+		if s2.Listener == nil {
+			panic("dnstest.NewServer(): failed to create tcp listener")
```
Comment (miekg) [inline plugin/pkg/dnstest/server.go]: shouldn't this be `continue`?

Comment (rdrozhdzh) [reply plugin/pkg/dnstest/server.go]: Here we ask system for **any** available port, so if it fails here this means that system has no available ports or some error not related to port. So, I don't think we should retry in this case

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] retry creatin..." ]
>rdrozhdzh commented on this pull request.
>
>
>
>> @@ -23,18 +23,28 @@ func NewServer(f dns.HandlerFunc) *Server {
> 	ch1 := make(chan bool)
> 	ch2 := make(chan bool)
>
>-	l, _ := net.Listen("tcp", ":0")
>-	if l == nil {
>-		return nil
>+	s1 := &dns.Server{} // udp
>+	s2 := &dns.Server{} // tcp
>+
>+	for i := 0; i < 1000; i++ { // 1000 attempts
>+		s2.Listener, _ = net.Listen("tcp", ":0")
>+		if s2.Listener == nil {
>+			panic("dnstest.NewServer(): failed to create tcp listener")
>
>Here we ask system for **any** available port, so if it fails here this means that system has no available ports or some error not related to port. So, I don't think we should retry in this case

Why not, we do vagely is similar thing for the udp port, as ports come and go, 
we can just as well retry again...

File: plugin/pkg/dnstest/server.go (hunk 1 above)
Comment (rdrozhdzh) [reply plugin/pkg/dnstest/server.go]: It depends on how much free port system has. Yes, 5 attempts may be enough. On the other hand I don't see any drawback with 1000 attempts - it won't take much time. If you prefer 5 I can change this

**Comment by rdrozhdzh** [comment]
> Why not, we do vagely is similar thing for the udp port, as ports come and go, we can just as well retry again...

I mean in case of TCP (first listener) the error is rather hypothetical and likely not related to free ports. In case of UDP (second listener) we request some specific port and the probability of failure for this specific port is quite high, though some other ports in the system is likely free.

I can change to continue if you insist

**Comment by miekg** [comment]
It's just a number, but 1000 feels like overkill, so pretty please :)

I would also favor removing the first panic and just try again for the tcp port allocation.

**Comment by rdrozhdzh** [comment]
Done :)

**Pull request merged**